ZEP-XXX-journal-record-format.md

```diff
@@ -0,0 +1,239 @@
+```
+
+A member has the following fields:
+type: 0 (INACTIVE) 1 (PASSIVE) 2 (PROMOTABLE) 3 (ACTIVE)
+updated : uint64
+memberId : string
+
+
+In summary, a record on the journal looks as follows:
+
+A record with an ApplicationEntry
+
+```
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Header |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | sbe_headers of RaftRecord | term | entrytype | sbe_headers of ApplicationEntry | lowestAsqn | highestAsqn | length | applicationData |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+```
+
+A record with a ConfigurationEntry
+```
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Header |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | sbe_headers of RaftRecord | term | entrytype | sbe_headers of ConfigurationEntry | timestamp | members |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+```
+
+A record with an InitialEntry
+```
++-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Header |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | sbe_headers of RaftRecord | term | entrytype |
++-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+```
+
+## Replication schema
+
+One approach is to replicate the serialized journal record as it is. However the raft module which sends and receives the replicated record needs to have access to raft term.
+Hence, the record needs to be (partially) deserialized at the raft layer.
+The serialization logic should be then shared by the raft and the journal.
+Currently the transport layer uses Kryo for serialization.
+So even if we replicate the journal record as it is, we are not getting much benefits from it.
+We would still need to copy the record before it is send.
+However, if we can make use of operating systems' feature to transfer bytes directly from the file to the network it would make sense to transfer the entire journal record as it is.
```

Comment:
I don't think we will get this any time soon (even with gRPC), but I think it's still laudable to reduce the number of copies when possible. So with gRPC we will be able to cut it down copying to 2 - once from file to the netty send buffer (which is direct memory, so still happening out-of-JVM), then once to the socket. Which in the end is still preferable to the current 4 - once to serializedRaftRecord, once to Kryo, once to Netty send buffer, once to socket.

Reply:
You can read more about the issue with FileRegion and gRPC here: https://github.com/grpc/grpc-java/issues/1054

Reply:
We can revisit it when we move to grpc. It is a breaking change anyway when we change the protocol, so we can also change this format.